Add fast path for ASCII in UTF-8 validation

bluss opened the pull request:
Add fast path for ASCII in UTF-8 validation

This speeds up the ASCII case (and long stretches of ASCII in otherwise
mixed UTF-8 data) when checking UTF-8 validity.

Benchmark results suggest that on purely ASCII input, we can improve
throughput (megabytes verified / second) by a factor of 13 to 14 (smallish input).
On XML and mostly English language input (en.wikipedia XML dump),
throughput improves by a factor 7 (large input).

On mostly non-ASCII input, performance increases slightly or is the
same.

The UTF-8 validation is rewritten to use indexed access; since all
access is preceded by a (mandatory for validation) length check, bounds
checks are statically elided by LLVM and this formulation is in fact the best
for performance. A previous version had losses due to slice to iterator
conversions.

A large credit to Björn Steinbrink who improved this patch immensely,
writing this second version.

Benchmark results on x86-64 (Sandy Bridge) compiled with -C opt-level=3.

Old code is `regular`, this PR is called `fast`.

Datasets:

- `ascii` is just ASCII (2.5 kB)
- `cyr` is cyrillic script with ascii spaces (5 kB)
- `dewik10` is 10MB of a de.wikipedia XML dump
- `enwik8` is 100MB of an en.wikipedia XML dump
- `jawik10` is 10MB of a ja.wikipedia XML dump

```
test from_utf8_ascii_fast        ... bench:         140 ns/iter (+/- 4) = 18221 MB/s
test from_utf8_ascii_regular     ... bench:       1,932 ns/iter (+/- 19) = 1320 MB/s
test from_utf8_cyr_fast          ... bench:      10,025 ns/iter (+/- 245) = 511 MB/s
test from_utf8_cyr_regular       ... bench:      10,944 ns/iter (+/- 795) = 468 MB/s
test from_utf8_dewik10_fast      ... bench:   6,017,909 ns/iter (+/- 105,755) = 1740 MB/s
test from_utf8_dewik10_regular   ... bench:  11,669,493 ns/iter (+/- 264,045) = 891 MB/s
test from_utf8_enwik8_fast       ... bench:  14,085,692 ns/iter (+/- 1,643,316) = 7000 MB/s
test from_utf8_enwik8_regular    ... bench:  93,657,410 ns/iter (+/- 5,353,353) = 1000 MB/s
test from_utf8_jawik10_fast      ... bench:  29,154,073 ns/iter (+/- 4,659,534) = 340 MB/s
test from_utf8_jawik10_regular   ... bench:  29,112,917 ns/iter (+/- 2,475,123) = 340 MB/s
```

Co-authored-by: Björn Steinbrink <bsteinbr@gmail.com>

rust-highfive commented:
r? @brson

(rust_highfive has picked a reviewer for you, use r? to override)

bluss commented:
Benchmarks using long texts are here: https://gist.github.com/bluss/bf45e07e711238e22b7a

2-3% slowdown on japanese and cyrillic texts that are mostly non-ascii. I don't have a problem championing that regression, given the speedup on utf-8 validation for predominantly ASCII input.

shepmaster commented on src/libcollectionstest/str.rs:
```diff
@@ -468,6 +468,18 @@ fn test_is_utf8() {
     assert!(from_utf8(&[0xF0, 0x90, 0x80, 0x80]).is_ok());
     assert!(from_utf8(&[0xF4, 0x8F, 0xBF, 0xBF]).is_ok());
+
+    // deny embedded in long stretches of ascii
```
I don't really know what this specific set of tests is doing.

I always have a bit of a sad when there are these giant "test everything" tests; my personal pref would be another test like `is_utf8_is_not_tricked_by_non_ascii_in_long_stretches_of_ascii`. No need to add `test_`, no need to have a comment, a failed test tells you what failed. :smile_cat:

bluss replied:
Entirely reasonable, no reason to share test name there, no common setup or anything. Fixed to have its own test function.

shepmaster commented on src/libcollectionstest/str.rs:
```diff
@@ -468,6 +468,18 @@ fn test_is_utf8() {
+    for i in 32..64 {
+        for elt in data.iter_mut() {
+            *elt = 0;
+        }
```
Just move the array into here, instead of rezeroing it?

```rust
for i in 32..64 {
    let mut data = [0; 128];
    // ...
}
```

bluss replied:
Nice! Fixed (amended commit).

shepmaster commented on src/libcore/str/mod.rs:
```diff
@@ -1074,26 +1075,82 @@ unsafe fn cmp_slice(a: &str, b: &str, len: usize) -> i32 {
+    debug_assert!(from <= v.len());
+    slice::from_raw_parts(v.as_ptr().offset(from as isize),
+                          v.len() - from)
+}
```
Huh. I always thought this method existed...

bluss replied:
It seems to exist for &str.

shepmaster replied:
Yeah, which I think is why I thought it existed for slices. Maybe a future PR :-)

shepmaster commented on src/libcore/str/mod.rs:
```diff
@@ -1074,26 +1075,82 @@ unsafe fn cmp_slice(a: &str, b: &str, len: usize) -> i32 {
+    let ptr = v.as_ptr();
+
+    let mut offset = 0;
+    if len >= 2 * usize::BYTES {
```
Why the `2`?

bluss replied:
The loop is unrolled by 2 (reads 2 usize per lap).

shepmaster replied:
Apologies, I wasn't very clear. I guess it's a two-part question:

* Why unroll at all?
* Why only unroll by 2?

bluss replied:
It's a bit arbitrary, I've only tried 1, 2, and 4 and compared performance. In the memchr code, where this is taken from it's to fill a 16-byte register on x86-64, but that doesn't happen here.

brson replied:
Can you extract the 2 to a const with a descriptive name about unrolling? Since I don't see any hand-unrolling here, I am guessing that the `if` statement allows the compiler to the unrolling according to the unrolling factor. This is not obvious to me. Can you also add a comment explaining?

bluss replied:
Yes, and the two `ptr.offset` and deref per iteration.

shepmaster commented:
Pedantically, I'd say it should be ASCII (all caps) as it's an acronym. Also `non_ascii` cause I'd normally write `non-ASCII`. All my comments are at your discretion to take or leave! :innocent:

ranma42 commented on src/libcore/str/mod.rs:
```diff
@@ -1074,26 +1075,82 @@ unsafe fn cmp_slice(a: &str, b: &str, len: usize) -> i32 {
+        }
+    }
+
+    // find the byte after the point the loop stopped
```
If the result of (x & 0x80808080_80808080) is non-zero, you can "immediately" find which byte it is using `leading_zeros() / 8`

bluss replied:
depends on endianness, it works fine with `.trailing_zeros()` on x86-64. It deserves to be tried for sure, but I can't make it be an improvement.

What llvm compiles the current code into, the beast it is, is actually `if contains_nonascii(u | v) { break; }` which seems to make for a much simpler computation inside the loop, and a tight loop.

I'm not 100% happy with the code in contains_nonascii, so any suggestion for improvement would be super welcome, feel free to take the code (from the benchmark link) and find something.

ranma42 replied:
I downloaded the gist, but I am having some trouble in getting the datasets you used. Specifically, I assumed that enwik8 should be http://mattmahoney.net/dc/enwik8.zip and that the specific version of the Japanese wiki should not matter much, but I have no idea about big10.

bluss replied:
yes, maybe you can just skip those datasets you don't have though? I could have provided everything better.

big10 is the dataset in http://vaskir.blogspot.ru/2015/09/regular-expressions-rust-vs-f.html

so it's the first 10MB of the unzipped file from https://drive.google.com/open?id=0B8HLQUKik9VtUWlOaHJPdG0xbnM

bluss replied:
jawik10 is the first 10MB from the unzip of http://dumps.wikimedia.org/archive/2006/2006-07/jawiki/20061016/jawiki-20061016-pages-articles.xml.bz2

brson commented:
Sweet wins. r=me but please do extract `2` to a more descriptive constant.

bluss commented:
Ok, I'll look over if there's a neat way to write the unrolling factor

bluss commented:
I received an improved version by @dotdash and it's an improvement you wouldn't believe.

- No slowdown for non-ascii cases (cyrillic test case improves for some reason)
- Less unsafe code
- Even faster on the pure ascii case. Ascii is now 14 times faster than before this PR. English wikipedia dump is 7 times faster than before.

Updated PR description & benchmarks are in there

@brson I addressed loop unrolling only by adding another comment for it, don't see a nice way to factor it out to a constant

Gankro commented:
Wow, awesome stuff!

bluss commented:
Pushed a fix, there was a missing conditional, let's try this in travis. I can't measure any difference in perf.

shepmaster commented:
As a bit of "real world" performance information, I pulled this down and used it for [SXD](https://github.com/shepmaster/sxd-document).

**Parsing a 16M XML file**

Valgrind reported that `str::from_utf8` took this much of the total run time:

Rust 1.5 | This PR
-----------|----------
5.47% | 0.29%

And I measured a ~1.25% overall speedup in the program.

**Parsing a 111M XML file**

Rust 1.5 | This PR
-----------|----------
4.12% | 0.22%

And I measured a ~1.1% overall speedup in the program.

----

Thanks for the awesome performance gains!

dotdash commented:
Thanks a lot @shepmaster! Always encouraging to get that kind of feedback! :heart_eyes_cat: And thanks to @bluss for getting this started, I've been completely blind to the masking quick check when I initially looked into this a few weeks ago! :beers:

bluss commented:
@shepmaster Awesome to see some numbers! I'm guessing your data files are almost purely ASCII (as a lot of the data in the world is).

@brson This is ready for re-review. It's the same algorithm, indexed access though, and the fast skip ahead loop is simpler, because it's only attempted at aligned locations. The main loop will progress to an aligned location quickly anyway, if the input is mostly ascii.

shepmaster commented:
> I'm guessing your data files are almost purely ASCII

Ah, yes, I meant to mention that. They indeed are pure-ASCII.

shepmaster commented on src/libcore/str/mod.rs:
```diff
@@ -1156,8 +1155,42 @@ fn run_utf8_validation_iterator(iter: &mut slice::Iter<u8>)
+                    }
+                }
+                // step from the point where the wordwise loop stopped
+                while offset < len && v[offset] < 128 {
```
Reading through this, I thought at first that `128` was another number relating to byte widths, then realized it is the ASCII cutoff value. Since this is also used above (`first >= 128`), perhaps another constant could be in order?

bluss replied:
hm, I don't think it's needed

shepmaster commented on src/libcore/str/mod.rs:
```diff
@@ -1074,46 +1075,44 @@ unsafe fn cmp_slice(a: &str, b: &str, len: usize) -> i32 {
 */
 
+// use truncation to fit u64 into usize
+const NONASCII_MASK: usize = 0x80808080_80808080u64 as usize;
```
Below, you chose to inline the constant inside the method; here you did not. Is there a method to these decisions?

bluss replied:
Didn't think much about it, but the second constant would be in the function by necessity so that it's not defined too far from its use.

shepmaster commented on src/libcore/str/mod.rs:
```diff
@@ -1156,8 +1155,42 @@ fn run_utf8_validation_iterator(iter: &mut slice::Iter<u8>)
+                                break;
+                            }
+                        }
+                        offset += UNROLLED_BY * usize::BYTES;
```
You could choose to introduce a constant like `const STRIDE_BYTES: usize = UNROLLED_BY * usize::BYTES` to avoid duplicating the math.

bluss commented:
I updated the second commit to use a constant for 2 * usize::BYTES instead, to follow shepmaster's suggestion roughly.

brson commented:
@bors r+

bors commented:
:pushpin: Commit cadcd70 has been approved by `brson`

<!-- @bors r=brson cadcd70775cf42b2add2526026a0a06c1ced411c -->

bors commented:
:hourglass: Testing commit cadcd70 with merge e7e4ecc...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/7705), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/7726), [auto-linux-64-debug-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/976), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/7663), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/7680), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/7665), [auto-linux-cross-opt](http://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/884), [auto-linux-musl-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/2835), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/7729), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/7725), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/7689), [auto-win-gnu-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/2709), [auto-win-gnu-32-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/2724), [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/2723), [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/2701), [auto-win-msvc-32-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/1943), [auto-win-msvc-64-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/2705)

bluss commented:
awesome. Thanks @brson and everyone.
